src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,284 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Framework.ConfigurationModel;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.Framework.OptionsModel
+{
+    public class ConfigurationBinder
+    {
+        private readonly Stack<IConfiguration> _context = new Stack<IConfiguration>();
+        private readonly IConfiguration _startConfiguration;
+
+        private IConfiguration _currentConfigurationContext;
+
+        public ConfigurationBinder([NotNull] IConfiguration configuration)
+        {
+            _startConfiguration = configuration;
+        }
+
+        public static void Bind(IConfiguration configuration, object merge)
+        {
+            new ConfigurationBinder(configuration).Bind(merge);
+        }
+
+        public void Bind(object model)
+        {
+            if (model == null)
+            {
+                return;
+            }
+
+            _context.Clear();
+            EnterContext(_startConfiguration);
+            BindObject(model);
+            ExitContext();
+        }
+
+        public TOptions Bind<TOptions>() where TOptions : new()
+        {
+            var model = new TOptions();
+            Bind(model);
+            return model;
+        }
+
+        private void BindObject(object obj)
+        {
+            foreach (var property in GetAllProperties(obj.GetType().GetTypeInfo()))
+            {
+                BindProperty(property, obj);
+            }
+        }
+
+        private void BindProperty(PropertyInfo property, object propertyOwner)
+        {
+            EnterContext(_currentConfigurationContext.GetSubKey(property.Name));
+            try
+            {
+                if (property.GetMethod == null || !property.GetMethod.IsPublic)
+                {
+                    // We don't support set only properties
+                    return;
+                }
+
+                var propertyValue = property.GetValue(propertyOwner);
+                var hasPublicSetter = property.SetMethod != null && property.SetMethod.IsPublic;
+
+                if (propertyValue == null && !hasPublicSetter)
+                {
+                    // Property doesn't have a value and we cannot set it so there is no
+                    // point in going further down the graph
+                    return;
+                }
+
+                propertyValue = BindType(property.PropertyType, propertyValue);
+
+                if (propertyValue != null && hasPublicSetter)
+                {
+                    property.SetValue(propertyOwner, propertyValue);
+                }
+            }
+            finally
+            {
+                ExitContext();
+            }
+        }
+
+
+        private object BindType(Type type, object typeInstance)
```

Comment:
It has nothing to do with properties. It just binds the current configuration node (and subnodes) to a type. If you pass an instance of that type it will populate that, if not it tries to create a new instance. It is really `BindType` :)